File: Packages/StackEngineCore/Tests/StackEngineCoreTests/ReferenceSelectionTests.swift
import XCTest
import simd
@testable import StackEngineCore

final class ReferenceSelectionTests: XCTestCase {
    private func checkerboard(_ n: Int) -> PixelImage {
        var img = PixelImage(width: n, height: n)
        for y in 0..<n { for x in 0..<n {
            let v: Float = ((x + y) % 2 == 0) ? 1 : 0
            img[x, y] = SIMD3<Float>(v, v, v)
        }}
        return img
    }
    private func flat(_ n: Int, _ v: Float) -> PixelImage {
        PixelImage(width: n, height: n, fill: SIMD3<Float>(v, v, v))
    }

    func testSharpnessHigherForCheckerboard() {
        XCTAssertGreaterThan(Luma.sharpness(checkerboard(8)), Luma.sharpness(flat(8, 0.5)))
    }
    func testReferenceSelectionPicksSharpest() {
        let frames = [flat(8, 0.5), checkerboard(8), flat(8, 0.3)]
        XCTAssertEqual(ReferenceSelection.sharpestIndex(frames), 1)
    }

    func testSingleFrameReturnsZero() {
        XCTAssertEqual(ReferenceSelection.sharpestIndex([flat(8, 0.5)]), 0)
    }
}

File: android/stackengine/src/test/kotlin/com/jdpneto/stackengine/ReferenceSelectionTests.kt
package com.jdpneto.stackengine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ReferenceSelectionTests {

    private fun checkerboard(n: Int): PixelImage {
        val img = PixelImage(n, n)
        for (y in 0 until n) for (x in 0 until n) {
            val v: Float = if ((x + y) % 2 == 0) 1f else 0f
            img[x, y] = Vec3(v, v, v)
        }
        return img
    }

    private fun flat(n: Int, v: Float): PixelImage = PixelImage(n, n, Vec3.repeating(v))

    @Test
    fun testSharpnessHigherForCheckerboard() {
        assertTrue(Luma.sharpness(checkerboard(8)) > Luma.sharpness(flat(8, 0.5f)))
    }

    @Test
    fun testReferenceSelectionPicksSharpest() {
        val frames = listOf(flat(8, 0.5f), checkerboard(8), flat(8, 0.3f))
        assertEquals(1, ReferenceSelection.sharpestIndex(frames))
    }

    @Test
    fun testSingleFrameReturnsZero() {
        assertEquals(0, ReferenceSelection.sharpestIndex(listOf(flat(8, 0.5f))))
    }
}
